src/drivers/uart/uart.c

```diff
@@ -83,13 +97,57 @@ w_status_t uart_init(uart_channel_t channel, UART_HandleTypeDef *huart, uint32_t
         return W_FAILURE;
     }
 
+    // Register the transmit-complete ISR for this UART channel
+    hal_status =
+        HAL_UART_RegisterCallback(huart, HAL_UART_TX_COMPLETE_CB_ID, HAL_UART_TxCpltCallback);
+    if (hal_status != HAL_OK) {
+        return W_FAILURE; // error occured in HAL_UART_RegisterCallback function
+    }
+
     // Start first reception
     if (HAL_UARTEx_ReceiveToIdle_IT(huart, handle->rx_msgs[0].data, UART_MAX_LEN) != HAL_OK) {
         return W_IO_ERROR;
     }
 
     return W_SUCCESS;
 }
+/**
+ * @brief Write to the specified UART channel
+ * @details // One task can write to a channel at once, and concurrent calls will block for 'timeout
+ * @param channel UART channel to write to
+ * @param data Buffer to store the data to send
+ * @param length Pointer to store the length of the sending message
+ * @param timeout_ms Timeout in milliseconds
+ * @return Status of the write operation
+ */
+w_status_t
+uart_write(uart_channel_t channel, uint8_t *buffer, uint8_t *length, uint32_t timeout_ms) {
+    w_status_t status = W_SUCCESS;
+    if ((channel >= UART_CHANNEL_COUNT) || (NULL == s_uart_handles[channel].huart)) {
+        status = W_INVALID_PARAM; // Invalid parameter(s)
+        return status;
+    } else if (pdTRUE != xSemaphoreTake(s_uart_handles[channel].write_mutex, timeout_ms)) {
+        return W_IO_TIMEOUT; // Could not acquire the mutex in the given time
+    }
+    HAL_StatusTypeDef transmit_status =
+        HAL_UART_Transmit_IT(s_uart_handles[channel].huart, buffer, *length);
+
+    if (HAL_ERROR == transmit_status) {
+        status = W_IO_ERROR;
+    } else if (HAL_TIMEOUT == transmit_status) {
+        status = W_IO_TIMEOUT;
+    }
```

Comment:
This isn't in the API description and I didn't check last time but looking at the [source code](https://github.com/waterloo-rocketry/cansw_processor_canards/blob/dad939f2d5f92fd0b0d819f657f68dedd50e3ab2/Drivers/STM32H7xx_HAL_Driver/Src/stm32h7xx_hal_uart.c#L1281-L1347) apparently `HAL_UART_Transmit_IT` will only ever return `HAL_OK`, `HAL_ERROR`, or `HAL_BUSY`, so that last condition would never be true. Should we be checking for `HAL_BUSY` instead of `HAL_TIMEOUT`?